### ISSUE TITLE
Read jacoco.xml from private gh-pages

### ISSUE THREAD

**Issue by jglick** [jenkins-x/ext-jacoco#50]
[This code](https://github.com/jenkins-x-apps/jx-app-jacoco/blob/64acef83070de85c16e7f404be03f7041b043c46/main.go#L108:L183) does not look like it would work when `jacoco.xml` was published to a private repository, as the GH raw URL will be inaccessible without an API token. I believe you need to use [this idiom](https://github.com/jenkins-x/jx/blob/b8f35c1fec60e359297cc506d40139ec15227631/pkg/jx/cmd/step_unstash.go#L101:L109), as alluded to in [storage documentation](https://jenkins-x.io/architecture/storage/#unstashing).

(It seems likely that there will be a lot of boilerplate code like this common to many apps processing build metadata. Perhaps some helper functions could be extracted to a common repository for reuse?)

**Comment by davidcurrie** [comment]
Jason made this change for `ext-go`: https://github.com/cloudbees/ext-go/commit/d97667c17160cdd9d305078fdfa7bf836e03e7ca

**Comment by fiunchinho** [comment]
is this solved then? can we close it?

**Comment by jglick** [comment]
What makes you think it is solved? Nothing has changed since I filed it.

**Comment by fiunchinho** [comment]
I thought I read a comment saying that this was fixed by a change in other PR. My bad then.

**Comment by hferentschik** [comment]
I seem to be missing a piece here, but how does `jacoco.xml` get onto gh-pages and what registers the URL in the pipeline activity? @fiunchinho @jglick can you shine some light on this?

**Comment by jglick** [comment]
> how does `jacoco.xml` get onto gh-pages

Currently, someone’s `Jenkinsfile` runs

```groovy
sh 'jx step collect -c jacoco -basedir target/site -p target/site/jacoco.xml'
```

or something along those lines. There was a plan for the app to somehow collect files according to a standard pattern automatically, according to a declarative aspect of the app which some infrastructure would interpret, but this is not implemented yet, maybe not even designed.

> what registers the URL in the pipeline activity?

`jx step collect` does this.

**Comment by hferentschik** [comment]
> Currently, someone’s `Jenkinsfile` runs

> ```groovy
> sh 'jx step collect -c jacoco -basedir target/site -p target/site/jacoco.xml'
> ```

Ahh, thanks that was indeed the piece I was missing. Makes sense now. It is actually also in the docs, but I did not connect the dots here. That said, I am wondering why we would use the imo rather confusing _collect_ alias at this point. The actual command is `jx step [stash|unstash]` which is ways more telling and would maybe have given me a notch into the right direction.

> There was a plan for the app to somehow collect files according to a standard pattern automatically, according to a declarative aspect of the app which some infrastructure would interpret, but this is not implemented yet, maybe not even designed.

Got you.

**Comment by hferentschik** [comment]
FYI, I created https://github.com/jenkins-x-apps/jx-app-jacoco/issues/74

**Comment by hferentschik** [comment]
> I believe you need to use [this idiom](https://github.com/jenkins-x/jx/blob/b8f35c1fec60e359297cc506d40139ec15227631/pkg/jx/cmd/step_unstash.go#L101:L109), as alluded to in [storage documentation](https://jenkins-x.io/architecture/storage/#unstashing).

I looked into it and I am not so sure whether this is going to work. This is using something like

```golang
authSvc, err := o.CreateGitAuthConfigService()
```

which iiuc looks up credentials which are stored in your local ~/.jx directory. This app is running in the cluster. Not sure how to get to the api token there. WDYT? Does this sounds right?

**Comment by jglick** [comment]
It just works. I do not know how.

**Comment by hferentschik** [comment]
> It just works. I do not know how.

I'll try. I actually already written the code.

**Comment by fiunchinho** [comment]
> I looked into it and I am not so sure whether this is going to work. This is using something like
> 
> ```go
> authSvc, err := o.CreateGitAuthConfigService()
> ```
> 
> which iiuc looks up credentials which are stored in your local ~/.jx directory. This app is running in the cluster. Not sure how to get to the api token there. WDYT? Does this sounds right?

I'm not sure how it works, but I thought running `jx step git credentials` as part of the pipeline was configuring git so that calls like that would work.

**Comment by jglick** [comment]
> running `jx step git credentials` as part of the pipeline

The issue here is that `ReadURL` needs to be called from an app, running in its own pod.

From experience with another similar app, I know that it does in fact work with a private repository.

**Comment by hferentschik** [comment]
Merged via pull request #97
